src/libAtomVM/mailbox.c

```diff
@@ -347,6 +356,26 @@ Message *mailbox_first(Mailbox *mbox)
     return result;
 }
 
+Message *mailbox_take(Mailbox *mbox)
+{
+    mailbox_reset(mbox);
+    MailboxMessage *msg = mbox->receive_pointer;
+    Message *first = NULL;
+    if (msg) {
+        first = &msg->body.normal;
+
+        mbox->inner_first = msg->header.next;
+        if (mbox->inner_first == NULL) {
+            // If this also the last, update inner_last.
+            mbox->inner_last = NULL;
+        }
+    }
+
+    mailbox_reset(mbox);
```

Comment:
Calling twice mailbox_reset sounds suspicious here, especially if msg is `NULL`.
You may want to simply update mbox->receive_pointer between lines 367 and 368.

Reply:
ok, I changed it:
```patch
@@ -365,14 +365,13 @@ Message *mailbox_take(Mailbox *mbox)
         first = &msg->body.normal;
 
         mbox->inner_first = msg->header.next;
+        mbox->receive_pointer = msg->header.next;
         if (mbox->inner_first == NULL) {
             // If this also the last, update inner_last.
             mbox->inner_last = NULL;
         }
     }
 
-    mailbox_reset(mbox);
-
     return first;
 }
```